src/main/java/com/bushidowallet/core/bitcoin/tx/input/MultiSigScriptHashInput.java

```diff
@@ -1,168 +1,205 @@
+        if (signatures != null) {
+            this.signatures = signatures;
+        } else {
+            this.signatures = new ArrayList<TransactionSignature>();
+        }
+        if (Script.buildScriptHashOut(this.redeemScript).equals(output.script) == false) {
+            throw new Exception("Provided public keys don't hash to the provided output");
+        }
+    }
+
+    public int countSignatures() {
+        if (signatures != null) {
+            return signatures.size();
+        }
+        return 0;
+    }
+
+    public int countMissingSignatures() {
+        return this.treshold - this.countSignatures();
+    }
+
+    @Override
+    public List<TransactionSignature> getSignatures(Transaction tx, ECKey key, int index, int sigType, byte[] hashData) throws Exception {
+        List<TransactionSignature> signatures = new ArrayList<TransactionSignature>();
+        for (ECKey publicKey : publicKeys) {
+            if (Arrays.equals(publicKey.getPublic(), key.getPublic())) {
+                TransactionSignature s = new TransactionSignature();
+                s.publicKey = key.getPublic();
+                s.prevTxId = this.prevTxId;
+                s.outputIndex = this.outputIndex;
+                s.inputIndex = index;
+                s.signature = SigHash.sign(tx, key, sigType, index, redeemScript);
+                s.sigType = sigType;
+                s.publicKeyCompressed = key.isCompressed();
+                signatures.add(s);
+            }
+        }
+        return signatures;
+    }
+
+    public void addSignature(Transaction tx, TransactionSignature signature) throws Exception {
+    	
+    	
+    	/** do NOT sign by liwl on 20161201 */
+    	if(tx != null){
+    		addSignatureByOne(tx, signature);
+    		return;
+    	}
+    	
+        if (isFullySigned() == false) {
+            if (this.publicKeyIndex.get(ByteUtil.toHex(signature.publicKey)) != null) {
+                if (isValidSignature(tx, signature)) {
+                    int index = this.publicKeyIndex.get(ByteUtil.toHex(signature.publicKey));
+                    if (signatures == null) {
+                        signatures = new ArrayList<TransactionSignature>();
+                    }
+                    signatures.add(index, signature);
+                    updateScript();
+                } else {
+                    throw new Exception("Attempting to add an invalid signature");
+                }
+            } else {
+                throw new Exception("Signature has no matching public key");
+            }
+        } else {
+            throw new Exception("All needed signatures have already been added");
+        }
+    }
+
+    /**
+     * add by liwl on 20161201
+     * add signature by one person.
+     * Different people in different machine.
+     * 
+     * @param tx
+     * @param signature
+     * @throws Exception
+     */
+    public void addSignatureByOne(Transaction tx, TransactionSignature signature) throws Exception {

```

Comment:
Besides, it is seems feasible to check if signature is correlated with public key used to generate a multisig address, don't You agree? Otherwise, it would be no way to validate if transaction is fully signed as invalid signatures would have been allowed in. I prefer strict API's, just like Bitpay's Bitcore Js authors. Renaming Your pull request to "Proposal to relax the API", but I am open for constructive conversations about the API.